# Homework #2 Part 1 - US Postal Service Zip Code

(This skeleton file is provided for HW#1 only.  You are expected to modify it for use in this and other homeworks for the course.)    

This is the README section for A0000000X's submission.
(For group submissions [when applicable], simply concatenate the student matric numbers in lexicographical order separated by a '-' (dash); e.g., A0000000X-A0000001Y)

### General Notes about this assignment 

Place your comments or requests here for the CS3244 staff to read.  Discuss your architecture or experiments in general.  A paragraph or two is usually sufficient.

### Files included with this submission

List the files in your submission here and provide a short 1 line description of each file.  Make sure your submission's files are named and formatted correctly.


## Programming Exercise

In [1]:
from sklearn import svm
import numpy as np
import numpy.random as nr
import matplotlib.pyplot as pl
%matplotlib inline

# Size the plot appropriately for online display
pl.rcParams['figure.figsize'] = (12.0, 10.0)

Let's fix the random number generator first, in case we need results that are replicable.

In [2]:
nr.seed(3244)

In [3]:
#load training dataset code
f = open('hw2-1-train.txt')
xn = [] # data input vector
yn = [] # data output vector

# Parse each line of training data file to get input and corresponding output(class) 
for line in f: 
    data = line.split()
    data = list(map(float,data))
    xi = data[1:len(data)]
    yi = data[0]
    xn.append(xi)
    yn.append(yi)
    
xn = np.asarray(xn) # convert input vector to numpy array
yn = np.asarray(yn) # convert output vector to numpy array
print(xn)
print(yn)

[[ 0.2479082  -4.1050625 ]
 [ 0.1595     -0.8375    ]
 [ 0.07964063 -0.79925   ]
 ..., 
 [ 0.13388281 -0.929875  ]
 [ 0.20566406 -0.5256875 ]
 [ 0.33945508 -5.1548125 ]]
[-1.  1.  1. ...,  1.  1. -1.]


In [4]:
#load testing dataset code
f = open('hw2-1-test.txt')
xntest = [] # data input vector
yntest = [] # data output vector

# Parse each line of testing data file to get input and corresponding output(class) 
for line in f: 
    data = line.split()
    data = list(map(float,data))
    xi = data[1:len(data)]
    yi = data[0]
    xntest.append(xi)
    yntest.append(yi)
    
xntest = np.asarray(xntest) # convert input vector to numpy array
yntest = np.asarray(yntest) # convert output vector to numpy array

In [5]:
# Train the SVM
def svmTrain(dataTrain, labelTrain, cost, kernel, gamma, degree):
    svmModel = svm.SVC(C=cost,coef0 = 1.0,degree=degree,gamma=gamma,kernel=kernel)
    svmModel.fit(dataTrain,labelTrain)
    totalSV = svmModel.support_vectors_
    return (svmModel,totalSV)

In [6]:
# Predict using SVM Model
def svmPredict(data,label,svmModel):
    predictedLabels = svmModel.predict(data)
    totalSamples = label.shape[0]
    '''
    testAccuracy = 0.0
    for i in range(totalSamples):
        if(labelTrain[i] != predictedLabels[i]):
            testAccuracy += 1.0
    testAccuracy = testAccuracy/totalSamples
    '''
    accuracy = svmModel.score(data,label)
    return accuracy

In [7]:
# Function to specify number of data points to be considered
def subset_data(data,labels,pointCount):
    return (data[0:pointCount,:],labels[0:pointCount])

In [8]:
# Problem 1A
def problem_1a(data, labels,pointCount):
    subset = subset_data(xn,yn,pointCount)
    xn_subset = subset[0]   # Subset of datapoints under consideration
    yn_subset = subset[1]   # Corresponding labels under consideration
    trainingResults = svmTrain(xn_subset,yn_subset,1.0,'linear','auto',3)
    svmModel = trainingResults[0];
    SV = trainingResults[1]
    totalSV = SV.shape[0]   # Total Support Vectors
    trainAccuracy = svmPredict(xn_subset,yn_subset,svmModel)
    testAccuracy = svmPredict(xntest,yntest,svmModel)
    #print('Train accuracy ' + str(pointCount) + '  '+str(trainAccuracy))
    print('Test accuracy for ' + str(pointCount) + ' points: '+str(testAccuracy))
    print('Number of support vectors for ' + str(pointCount) + ' points: '+str(totalSV))
    print()
    return ('linear','1.0',totalSV,trainAccuracy,testAccuracy)
    
def write_results_to_file(results,mode):
    fo = open('problem-1a.txt', mode)
    kernel = results[0]
    cost = results[1]
    totalSV = results[2]
    trainAccuracy = results[3]
    testAccuracy = results[4]
    fo.write("Kernel: "+ str(kernel)+"\n")
    fo.write("Cost: "+ str(cost)+ "\n")
    fo.write("Number of Support Vectors: "+ str(totalSV)+"\n")
    fo.write("Train Accuracy: "+ str(trainAccuracy)+"\n")
    fo.write("Test Accuracy: " + str(testAccuracy)+"\n")
    fo.write("\n")
    fo.close()
    
# Evaluate on all points
results = problem_1a(xn,yn,xn.shape[0])
write_results_to_file(results,'w')

# Evaluate on 50 points
results = problem_1a(xn,yn,50)
#write_results_to_file(results,'a')

# Evaluate on 100 points
results = problem_1a(xn,yn,100)
#write_results_to_file(results,'a')

# Evaluate on 200 points
results = problem_1a(xn,yn,200)
#write_results_to_file(results,'a')

# Evaluate on 800 points
results = problem_1a(xn,yn,800)
#write_results_to_file(results,'a')


Test accuracy for 1561 points: 0.978773584906
Number of support vectors for 1561 points: 28

Test accuracy for 50 points: 0.971698113208
Number of support vectors for 50 points: 2

Test accuracy for 100 points: 0.978773584906
Number of support vectors for 100 points: 4

Test accuracy for 200 points: 0.981132075472
Number of support vectors for 200 points: 4

Test accuracy for 800 points: 0.978773584906
Number of support vectors for 800 points: 10



Answer to problem-1a ( file output given ONLY for entire dataset as instructed )

Test accuracy for 1561 points: 0.978773584906 <br>
Number of support vectors for 1561 points: 28 <br>

Test accuracy for 50 points: 0.971698113208 <br>
Number of support vectors for 50 points: 2 <br>

Test accuracy for 100 points: 0.978773584906 <br>
Number of support vectors for 100 points: 4 <br>

Test accuracy for 200 points: 0.981132075472 <br>
Number of support vectors for 200 points: 4 <br>

Test accuracy for 800 points: 0.978773584906 <br>
Number of support vectors for 800 points: 10

In [9]:
# Problem 1B
def problem_1b(data, labels,degree,cost):
    gamma = 1.0
    trainingResults = svmTrain(data,labels,cost,'poly',gamma,degree)
    svmModel = trainingResults[0];
    SV = trainingResults[1]
    totalSV = SV.shape[0]
    trainAccuracy = svmPredict(xn,yn,svmModel)
    testAccuracy = svmPredict(xntest,yntest,svmModel)
    if(cost == 0.001):
        print('Total support vectors at Q=' +str(degree) + ':  '+str(totalSV))
    elif(cost == 1):
        print('Eout at Q=' +str(degree) + ':  '+str(1-testAccuracy))
    else:
        print('Ein at Q=' +str(degree) + ':  '+str(1-trainAccuracy))
    #print('Eout ' + str(cost) + '  '+str(degree) + '  '+str(1-testAccuracy))
    return ('polynomial',cost,totalSV,trainAccuracy,testAccuracy)

def write_results_to_file_1b(filename,results,mode,degree):
    fo = open(filename, mode)
    kernel = results[0]
    cost = results[1]
    totalSV = results[2]
    trainAccuracy = results[3]
    testAccuracy = results[4]
    fo.write("Kernel: "+ str(kernel)+"\n")
    fo.write("Degree (Q): "+str(degree)+"\n")
    fo.write("Cost: "+ str(cost)+ "\n")
    fo.write("Number of Support Vectors: "+ str(totalSV)+"\n")
    fo.write("Train Accuracy: "+ str(trainAccuracy)+"\n")
    fo.write("Test Accuracy: " + str(testAccuracy)+"\n")
    fo.write("\n")
    fo.close()
# Q = 2 or Q = 5

# Part (i) [ C = 0.0001]
cost = 0.0001
answer_1 = False
results_q2 = problem_1b(xn,yn,2,cost)
results_q5 = problem_1b(xn,yn,5,cost)
Ein_q2 = 1 - results_q2[3]
Ein_q5 = 1 - results_q5[3]
if(Ein_q5 > Ein_q2):
    print('Statement (i) is true')
    answer_1 = True
else:
    print('Statement (i) is false')
print()
write_results_to_file_1b('problem-1b.txt',results_q2,'w',2)
write_results_to_file_1b('problem-1b.txt',results_q5,'a',5)

# Part (ii) [ C = 0.001]
cost = 0.001
answer_2 = False
results_q2 = problem_1b(xn,yn,2,cost)
results_q5 = problem_1b(xn,yn,5,cost)
totalSV_q2 = results_q2[2]
totalSV_q5 = results_q5[2]
if(totalSV_q5 < totalSV_q2):
    print('Statement (ii) is true')
    answer_2 = True
else:
    print('Statement (ii) is false')
print()
write_results_to_file_1b('problem-1b.txt',results_q2,'a',2)
write_results_to_file_1b('problem-1b.txt',results_q5,'a',5)
    
# Part (iii) [ C = 0.01]
cost = 0.01
answer_3 = False
results_q2 = problem_1b(xn,yn,2,cost)
results_q5 = problem_1b(xn,yn,5,cost)
Ein_q2 = 1 - results_q2[3]
Ein_q5 = 1 - results_q5[3]
if(Ein_q5 > Ein_q2):
    print('Statement (iii) is true')
    answer_3 = True
else:
    print('Statement (iii) is false')
print()
write_results_to_file_1b('problem-1b.txt',results_q2,'a',2)
write_results_to_file_1b('problem-1b.txt',results_q5,'a',5)
    
# Part (iv) [ C = 1]
cost = 1
answer_4 = False
results_q2 = problem_1b(xn,yn,2,cost)
results_q5 = problem_1b(xn,yn,5,cost)
Eout_q2 = 1 - results_q2[4]
Eout_q5 = 1 - results_q5[4]
if(Eout_q5 < Eout_q2):
    print('Statement (iv) is true')
    answer_4 = True
else:
    print('Statement (iv) is false')
print()
write_results_to_file_1b('problem-1b.txt',results_q2,'a',2)
write_results_to_file_1b('problem-1b.txt',results_q5,'a',5)
    
# Part (v)
if(not(answer_1 or answer_2 or answer_3 or answer_4)):
    print('Statement (v) is true')
else:
    print('Statement (v) is false')
print()

Ein at Q=2:  0.00896860986547
Ein at Q=5:  0.00448430493274
Statement (i) is false

Total support vectors at Q=2:  76
Total support vectors at Q=5:  25
Statement (ii) is true

Ein at Q=2:  0.00448430493274
Ein at Q=5:  0.00384368994234
Statement (iii) is false

Eout at Q=2:  0.0188679245283
Eout at Q=5:  0.0212264150943
Statement (iv) is false

Statement (v) is false



Answer to problem-1b ( answers also given in .txt file for all cases)

Ein at Q=2:  0.00896860986547 <br>
Ein at Q=5:  0.00448430493274 <br>
Statement (i) is false <br>

Total support vectors at Q=2:  76 <br>
Total support vectors at Q=5:  25 <br>
Statement (ii) is true <br>

Ein at Q=2:  0.00448430493274 <br>
Ein at Q=5:  0.00384368994234 <br>
Statement (iii) is false <br>

Eout at Q=2:  0.0188679245283 <br>
Eout at Q=5:  0.0212264150943 <br>
Statement (iv) is false <br>

Statement (v) is false

In [10]:
# Problem 1c

def problem_1c(dataTrain, labelsTrain, dataTest, labelsTest,cost):
    kernel = 'rbf'
    gamma = 1.0
    degree = 1
    trainingResults = svmTrain(dataTrain,labelsTrain,cost,kernel,gamma,degree)
    svmModel = trainingResults[0];
    SV = trainingResults[1]
    totalSV = SV.shape[0]
    trainAccuracy = svmPredict(dataTrain,labelsTrain,svmModel)
    testAccuracy = svmPredict(dataTest,labelsTest,svmModel)
    print('Ein for cost ' + str(cost) +' is: '+str(1-trainAccuracy))
    print('Eout for cost ' + str(cost) + ' is: '+str(1-testAccuracy))
    print()
    return ('polynomial',cost,totalSV,trainAccuracy,testAccuracy)

def write_results_to_file_1c(filename,results,mode):
    fo = open(filename, mode)
    kernel = results[0]
    cost = results[1]
    totalSV = results[2]
    trainAccuracy = results[3]
    testAccuracy = results[4]
    fo.write("Kernel: "+ str(kernel)+"\n")
    fo.write("Cost: "+ str(cost)+ "\n")
    fo.write("Number of Support Vectors: "+ str(totalSV)+"\n")
    fo.write("Train Accuracy: "+ str(trainAccuracy)+"\n")
    fo.write("Test Accuracy: " + str(testAccuracy)+"\n")
    fo.write("\n")
    fo.close()
    
overall_results_ein = []
overall_results_eout = []

# [ C = 0.01]
cost = 0.01
results_1 = problem_1c(xn,yn,xntest,yntest,cost)
Ein_1 = 1 - results_1[3]
Eout_1 = 1 - results_1[4]
overall_results_ein.append((cost,Ein_1))
overall_results_eout.append((cost,Eout_1))
write_results_to_file_1c('problem-1c.txt',results_1,'w')

# [ C = 1 ]
cost = 1
results_2 = problem_1c(xn,yn,xntest,yntest,cost)
Ein_2 = 1 - results_2[3]
Eout_2 = 1 - results_2[4]
overall_results_ein.append((cost,Ein_2))
overall_results_eout.append((cost,Eout_2))
write_results_to_file_1c('problem-1c.txt',results_2,'a')

# [ C = 100 ]
cost = 100
results_3 = problem_1c(xn,yn,xntest,yntest,cost)
Ein_3 = 1 - results_3[3]
Eout_3 = 1 - results_3[4]
overall_results_ein.append((cost,Ein_3))
overall_results_eout.append((cost,Eout_3))
write_results_to_file_1c('problem-1c.txt',results_3,'a')

# [ C = 10000 ]
cost = 10000
results_4 = problem_1c(xn,yn,xntest,yntest,cost)
Ein_4 = 1 - results_4[3]
Eout_4 = 1 - results_4[4]
overall_results_ein.append((cost,Ein_4))
overall_results_eout.append((cost,Eout_4))
write_results_to_file_1c('problem-1c.txt',results_4,'a')

# [ C = 1000000 ]
cost = 1000000
results_5 = problem_1c(xn,yn,xntest,yntest,cost)
Ein_5 = 1 - results_5[3]
Eout_5 = 1 - results_5[4]
overall_results_ein.append((cost,Ein_5))
overall_results_eout.append((cost,Eout_5))
write_results_to_file_1c('problem-1c.txt',results_5,'a')

min_ein = min(overall_results_ein, key = lambda t: t[1])
min_eout = min(overall_results_eout, key = lambda t: t[1])
print()
print('The lowest value of Ein is '+str(min_ein[1])+' at C='+str(min_ein[0]))
print('The lowest value of Eout is '+str(min_eout[1])+' at C='+str(min_eout[0]))

Ein for cost 0.01 is: 0.00384368994234
Eout for cost 0.01 is: 0.0235849056604

Ein for cost 1 is: 0.00448430493274
Eout for cost 1 is: 0.0212264150943

Ein for cost 100 is: 0.00320307495195
Eout for cost 100 is: 0.0188679245283

Ein for cost 10000 is: 0.00256245996156
Eout for cost 10000 is: 0.0235849056604

Ein for cost 1000000 is: 0.000640614990391
Eout for cost 1000000 is: 0.0235849056604


The lowest value of Ein is 0.000640614990391 at C=1000000
The lowest value of Eout is 0.0188679245283 at C=100


Answer to problem-1c ( answers also given in .txt file for all subsets)

Ein for cost 0.01 is: 0.00384368994234 <br>
Eout for cost 0.01 is: 0.0235849056604

Ein for cost 1 is: 0.00448430493274 <br>
Eout for cost 1 is: 0.0212264150943

Ein for cost 100 is: 0.00320307495195 <br>
Eout for cost 100 is: 0.0188679245283

Ein for cost 10000 is: 0.00256245996156 <br>
Eout for cost 10000 is: 0.0235849056604

Ein for cost 1000000 is: 0.000640614990391 <br>
Eout for cost 1000000 is: 0.0235849056604


The lowest value of Ein is 0.000640614990391 at C=1000000 <br>
The lowest value of Eout is 0.0188679245283 at C=100

## Statement of Individual Work

Please initial (between the square brackets) one of the following statements.

[ MG ] I, A0153196B, certify that I have followed the CS 3244 Machine Learning class guidelines for homework assignments.  In particular, I expressly vow that I have followed the Facebook rule in discussing with others in doing the assignment and did not take notes (digital or printed) from the discussions.  

[ ] I, <*substitute your matric number here*>, did not follow the class rules regarding the homework assignment, because of the following reason:

<*Please fill in*>

I suggest that I should be graded as follows:

<*Please fill in*>

### References

I have refered to the following list of people and websites in preparing my homework submission:

<*please fill in*>